ceno_rt/src/mmio.rs

```diff
@@ -0,0 +1,53 @@
+//! Memory-mapped I/O (MMIO) functions.
+
+use rkyv::{Portable, api::high::HighValidator, bytecheck::CheckBytes, rancor::Failure};
+
+use core::slice::from_raw_parts;
+
+/// The memory region with our hints.
+///
+/// Logically, this is a static constant, but the tytpe system doesn't see it that way.
+/// (We hope that the optimiser is smart enough to see that it is a constant.)
+fn hints_region<'a>() -> &'a [u8] {
+    extern "C" {
+        /// The address of this variable is the start of the hints ROM.
+        ///
+        /// It is defined in the linker script.  The value of this variable is undefined.
+        static _hints_start: u8;
+        /// The _address_ of this variable is the length of the hints ROM
+        ///
+        /// It is defined in the linker script.  The value of this variable is undefined.
+        static _hints_length: usize;
+    }
+    unsafe {
+        let hints_length = &_hints_length as *const usize as usize;
+        from_raw_parts(&_hints_start, hints_length)
+    }
+}
+
+/// Get the length of the next hint
+fn hint_len() -> usize {
+    extern "C" {
+        /// The address of this variable is the start of the slice that holds the length of the hints.
+        ///
+        /// It is defined in the linker script.  The value of this variable is undefined.
+        static _lengths_of_hints_start: usize;
+    }
+    static mut NEXT_HINT_LEN_AT: *const usize = &raw const _lengths_of_hints_start;
+    unsafe {
+        let len: usize = core::ptr::read(NEXT_HINT_LEN_AT);
+        NEXT_HINT_LEN_AT = NEXT_HINT_LEN_AT.add(1);
+        len
+    }
+}
+
+pub fn read_slice<'a>() -> &'a [u8] {
+    &hints_region()[..hint_len()]
```

Comment:
Please see the [comment](https://github.com/scroll-tech/ceno/pull/631/files#r1886274179) that's trying to explain this.

> say the hints_region() is static constant, be [2,3,4,5,67,14,65] (if we don't talk about alignment for simplicity).
> 
> lens is [2, 3,..]
> 
> 1st time, read_slice will return [2,3], first 2 elements.
> 2nd time, read_slice will return [4,5,67] or [2,3,4]?

It would return [2,3,4].

> my understanding is that it should return [4,5,67], but if hints_region() is the full hints bytes, it will return [2,3,4]?

Yes, exactly, it'll return [2, 3, 4] and that's as intended.

We always return from the very start of the hints region.  I added a comment in the code (linked above) to explain this:

> Our guest programs have two requirements on the format:
> 1. The start of the hints buffer consists of a sequence of `usize` values, each representing the
>    length of the next hint (from the start of the whole buffer).
> 2. `hints[..current_hint_len]` can deserialise into the expected type via rkyv.
> 
> Note how we overlap the two areas, and don't specify starts for our hints.  That's a simplification
> and performance improvement we can make because of how rkyv works: you can add arbitrary padding to
> the left of a serialised buffer, and it will still work.

Let me know if it makes sense.